nav2_util/include/nav2_util/simple_action_server.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_UTIL__SIMPLE_ACTION_SERVER_HPP_
+#define NAV2_UTIL__SIMPLE_ACTION_SERVER_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_action/rclcpp_action.hpp"
+
+namespace nav2_util
+{
+
+template<typename ActionT>
+class SimpleActionServer
+{
+public:
+  typedef std::function<
+      void (const std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>>)> ExecuteCallback;
+
+  explicit SimpleActionServer(
+    rclcpp::Node::SharedPtr node,
+    const std::string & action_name,
+    ExecuteCallback execute_callback)
+  : node_(node), execute_callback_(execute_callback)
+  {
+    auto handle_goal =
+      [](const rclcpp_action::GoalUUID &, std::shared_ptr<const typename ActionT::Goal>)
+      {
+        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+      };
+
+    auto handle_cancel = [](std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>>)
+      {
+        return rclcpp_action::CancelResponse::ACCEPT;
+      };
+
+    auto handle_accepted = [this](std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>> handle)
+      {
+        // If we're currently working on a task, set a flag so that the
+        // action server can grab the pre-empting request in its loop
+        if (received_handle_ != nullptr && received_handle_->is_active()) {
```

Comment:
The server_goal_handle class itself serializes access with its own mutex: https://github.com/ros2/rclcpp/blob/master/rclcpp_action/src/server_goal_handle.cpp, which is a better place for it, IMO, so that clients of the class don't have to.

Reply:
It only serializes access to the server goal handle members. You've added additional members in this class that also need to be synced.

Reply:
Ok, true. I've added a mutex to serialize access.